packages/react-mongo/react-mongo.ts

```diff
@@ -0,0 +1,206 @@
+
+type useFindActions<T> =
+  | { type: 'refresh', data: T[] }
+  | { type: 'addedAt', document: T, atIndex: number }
+  | { type: 'changedAt', document: T, atIndex: number }
+  | { type: 'removedAt', atIndex: number }
+  | { type: 'movedTo', fromIndex: number, toIndex: number }
+
+const useFindReducer = <T>(data: T[], action: useFindActions<T>): T[] => {
+  switch (action.type) {
+    case 'refresh':
+      return action.data
+    case 'addedAt':
+      return [
+        ...data.slice(0, action.atIndex),
+        action.document,
+        ...data.slice(action.atIndex)
+      ]
+    case 'changedAt':
+      return [
+        ...data.slice(0, action.atIndex),
+        action.document,
+        ...data.slice(action.atIndex + 1)
+      ]
+    case 'removedAt':
+      return [
+        ...data.slice(0, action.atIndex),
+        ...data.slice(action.atIndex + 1)
+      ]
+    case 'movedTo':
+      const doc = data[action.fromIndex]
+      const copy = [
+        ...data.slice(0, action.fromIndex),
+        ...data.slice(action.fromIndex + 1)
+      ]
+      copy.splice(action.toIndex, 0, doc)
+      return copy
+  }
+}
+
+const useFindClient = <T = any>(factory: () => Mongo.Cursor<T>, deps: DependencyList) => {
+  const [data, dispatch] = useReducer<Reducer<T[], useFindActions<T>>, T[]>(
+    useFindReducer,
+    undefined,
+    () => Tracker.nonreactive(() => {
+      const cursor = factory()
+      return cursor.fetch()
+    })
+  )
+
+  useEffect(() => {
+    // Rebuild the cursor and data in case an update may have
+    // happened between first render and commit. Alternatively,
+    // update in response to deps change.
+    const [cursor, data] = Tracker.nonreactive(() => {
+      const cursor = factory()
+      return [ cursor, cursor.fetch() ]
+    })
+    dispatch({
+      type: 'refresh',
+      data: data
+    })
+
+    const observer = cursor.observe({
+      addedAt (document, atIndex, before) {
+        dispatch({ type: 'addedAt', document, atIndex })
+      },
+      changedAt (newDocument, oldDocument, atIndex) {
+        dispatch({ type: 'changedAt', document: newDocument, atIndex })
+      },
+      removedAt (oldDocument, atIndex) {
+        dispatch({ type: 'removedAt', atIndex })
+      },
+      movedTo (document, fromIndex, toIndex, before) {
+        dispatch({ type: 'movedTo', fromIndex, toIndex })
+      },
+      // @ts-ignore
+      _suppress_initial: true
+    })
+
+    return () => {
+      observer.stop()
+    }
+  }, deps)
+
+  return data
+}
+
+const useFindServer = <T = any>(factory: () => Mongo.Cursor<T>, deps: DependencyList) => Tracker.nonreactive(() => factory().fetch())
+
+export const useFind = Meteor.isServer
+  ? useFindServer
+  : useFindClient
+
+const useTrackerClient = <T = unknown>(reactiveFn: () => T, deps: DependencyList): T => {
+  const [data, setData] = useState(() => Tracker.nonreactive(reactiveFn))
+
+  useEffect(() => {
+    const computation = Tracker.nonreactive(() => (
```

Comment:
Why do we wrap the autorun with Tracker.nonreactive? Might be worth writing a comment here if it's required.

Reply:
This a common pattern in React apps, this is always needed to prevent the computation from being rerun on every render I think. There's a comment somewhere explaining it, I'll see if I can find it.

Reply:
Hmm, I wonder if it's needed in this case, I would have thought it's pretty safe as it's inside of a `useEffect`.

Reply:
It's just safe, I suppose, for that example. Is that the only condition? How common is it to see someone calling `ReacDOM.render` from inside of a computation?

Reply:
I don't know if that comment is completely accurate. I found that I needed it [in my package](https://github.com/rijk/meteor-subscribe-suspense/blob/main/src/use-subscription.ts#L21-L29) as well, otherwise the subscription would be stopped unexpectedly by the invisible parent Tracker that apparently wraps the whole app and runs on every render.

However, this is _not_ inside a `useEffect` but a synchronous call inside a render. So this might be different in this case as you suggested.

Reply:
I think we just need to confirm that there is a definitive purpose to it in this context. The only scenario where I can think React might be wrapped inside of an outer computation is if using [react-template-helper](https://guide.meteor.com/react.html#react-in-blaze)

Reply:
I'd be curious for the answer as well. I can tell you I'm not using any of that stuff, just plain React 17 with hooks, and there still was a hidden 'outer' Tracker computation that ran on renders.

Reply:
There's a package out there that makes render methods reactive by default. I also don't know if we need it in useEffect - it's defensive coding in that case. It's relatively cheap to do that, so I didn't worry about it. I could add that comment back for the first invocation.

Reply:
It's not correct to cover a hypothetical condition. We should prove it to be valuable and if it's not then just remove it.

Reply:
I agree, but time is limited! We still need tests and readmes, and a bunch of other stuff. Maybe we can add a test or some tests to figure this out?

Reply:
I can't find definitive documentation on how `useEffect` works - I'd need a guarantee that it'll never run synchronously with render at the least, but probably also a description of the context in which it runs. My guess is it always runs at least a tick after the current render, but that's just a guess. I also think semantically, it makes sense to wrap that in nonreactive blocks, because we are deliberately opting out of using Tracker in these cases. And again, it's super cheap - it doesn't cost much (or might even be a super tiny net gain) in terms of performance.

Reply:
I took it out, and it passed the tests. Let's leave it like that for now.